### PR TITLE
Forum support

### LINKED ISSUE #1372: Add support for forums
# Summary
Exactly what it says on the tin.

# Details
The [`ChannelType`](https://github.com/DSharpPlus/DSharpPlus/blob/master/DSharpPlus/Enums/Channel/ChannelType.cs) enum does not currently list `GUILD_FORUM` (type 15) as a supported value (it also does not list `GUILD_DIRECTORY` - type 14 - but that is unrelated to the issue at hand here)

As per the [API reference](https://discord.com/developers/docs/resources/channel#channel-object-channel-types), Forums are a new feature that are still being rolled out and innovated, but this should still be on the roadmap for D#+ imo. It appears that they are just the same as any other channel object, the only difference being that you can only create threads in them (i.e. "posts") - not send messages directly to them.

# Notes
The endpoint for creating a thread inside a forum is the same as creating the thread anywhere else (see https://discord.com/developers/docs/resources/channel#start-thread-in-forum-channel), but it appears to accept different parameters in the body. Namely `invitable` and `type` are not accepted, but a `message` is required.

### PR DESCRIPTION
**Pull request by VelvetThePanda**
The PR you've all been waiting for. (Closes #1372)

### CODE REVIEW

File: DSharpPlus/Entities/Channel/Thread/DiscordThreadChannel.cs [hunk 1]
```diff
@@ -71,6 +72,18 @@ public class DiscordThreadChannel : DiscordChannel
...
         public bool IsNew { get; internal set; }
 
+        /// <summary>
+        /// Gets the tags applied to this forum post.
+        /// </summary>
+        // Performant? No. Ideally, you're not using this property often.
+        public IReadOnlyList<DiscordForumTag> AppliedTags =>
+        this.Parent is DiscordForumChannel parent
+        ? parent.AvailableTags.Where(pt => _appliedTagIds.Contains(pt.Id)).ToArray()
+        : Array.Empty<DiscordForumTag>();
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/Thread/DiscordThreadChannel.cs:82]: formatting - can we indent this by a level

File: DSharpPlus/Entities/Channel/Thread/Forum/DiscordForumChannel.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        public DefaultReaction? DefaultReaction { get; internal set; }
+
+        [JsonProperty("default_sort_order", NullValueHandling = NullValueHandling.Ignore)]
+        public DefaultSortOrder? DefaultSortOrder { get; internal set; }
+
+        [JsonProperty("default_forum_layout", NullValueHandling = NullValueHandling.Ignore)]
+        public DefaultForumLayout? DefaultForumLayout { get; internal set; }
+
+        public Task<DiscordForumPostStarter> CreateForumPostAsync(ForumPostBuilder builder)
+            => this.Discord.ApiClient.CreateForumPostAsync(this.Id, builder.Name, builder.AutoArchiveDuration, builder.SlowMode, builder.Message, builder.AppliedTags.Select(t => t.Id));
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/Thread/Forum/DiscordForumChannel.cs:69]: docs

File: DSharpPlus/Entities/Channel/Thread/Forum/ForumPostBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace DSharpPlus.Entities
+{
+    public class ForumPostBuilder
+    {
+        public string Name { get; set; }
+        public int? SlowMode { get; set; }
+        public DiscordMessageBuilder Message { get; set; }
+        public IReadOnlyList<DiscordForumTag> AppliedTags { get; }
+        public AutoArchiveDuration? AutoArchiveDuration { get; set; }
```
Comment (akiraveliara) [inline DSharpPlus/Entities/Channel/Thread/Forum/ForumPostBuilder.cs:35]: potentially document these? at the very least the class should have a xmldoc

File: DSharpPlus/Enums/Channel/Forum/ChannelFlags.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+namespace DSharpPlus
+{
+    public enum ChannelFlags
```
Comment (akiraveliara) [inline DSharpPlus/Enums/Channel/Forum/ChannelFlags.cs:25]: mark this as a flags enum, perhaps?

**Review by akiraveliara — COMMENTED** [review]
all in all it looks good at a glance

**Comment by R3-BTC** [comment]
when merge please?

**Comment by akiraveliara** [comment]
whenever we're sure it works. feel free to mess around with the PR and report any issues tou find.

**Comment by R3-BTC** [comment]
> whenever we're sure it works. feel free to mess around with the PR and report any issues tou find.

yeah i did, i found nothing that caused issues

**Comment by OoLunar** [comment]
We found a few issues in https://discord.com/channels/379378609942560770/1065106184316649584, not merging until those have been fixed

**Comment by VelvetThePanda** [comment]
Seen; will fix later if InF doesn't first

**Comment by VelvetThePanda** [comment]
Well, time to :shipit:. Thanks to everyone who's tested!

**Pull request merged**